### DISCUSSION TITLE
Rendering id as the exact string of the node rather than outputting to kebab-case

### DISCUSSION THREAD

**Discussion by cryptobench** [markdoc/markdoc#395] [Q&A] [answered]
I'm currently engaged in developing custom documentation software, during which I have been exploring the functionalities provided by Markdoc. However, I have encountered an issue pertaining to the way Markdoc generates HTML ID attributes which I would appreciate your assistance with.

In the context of my project, I am using TypeDoc to generate markdown files for a specific library. One such output is as follows:

```md
### Properties

- [id](Activity#id)
- [agreementId](Activity#agreementid)
Here, it links to the ID agreementid.
```

And for this I have created a node inside markdoc

```typescript
  h3: {
    ...defaultNodes.h3,
    attributes: {
      id: {
        type: String,
      },
    },
  },

```
The above snippet is expected to assign the h3's name to an ID, and as per the previously mentioned TypeDoc example, I was anticipating it to be `agreementid`. However, the rendered HTML instead provides `agreement-id`.

<img width="326" alt="image" src="https://github.com/markdoc/markdoc/assets/33448819/5c328c04-5fbc-4851-91ae-8f3aa89a6ceb">

This seems to suggest that Markdoc is converting the ID into kebab-case. I am seeking a way to configure Markdoc such that the HTML ID matches the exact h3 string, without this automatic conversion.

I would greatly appreciate any insights or suggestions you could provide on how to address this issue. 

Thank you for your time and assistance!

**Comment by rpaul-stripe** [comment]
Markdoc does not automatically add heading IDs by default and it does not have an `h3` node type—it has a single `heading` node type with a `level` attribute that specifies the heading level. The `id` attribute is also implicit on all Markdoc nodes, so it's generally not declared in a schema.

Edit: I think this is coming from the [Markdoc starter](https://github.com/markdoc/markdoc-starter/blob/main/markdoc/nodes/heading.markdoc.ts#L5). If you want to change the logic that is used to generate the ID, just modify the `generateID` function in the `markdoc/nodes/heading.markdoc.ts` file. I believe you can get the result you want by removing this line: `.replace(/\s+/g, '-')`

**Reply by cryptobench** [reply]
@rpaul-stripe I had no config for heading in my nodes already but I took the one from the example you provided, and it appears to fix my issue **but** the sidebar I had on the right with links completely disappears and doesn't render anything now


```js
function generateID(children, attributes) {
  if (attributes.id && typeof attributes.id === 'string') {
    return attributes.id
  }
  return children
    .filter((child) => typeof child === 'string')
    .join(' ')
    .replace(/[?]/g, '')
    .replace(/\s+/g, '-')
    .toLowerCase()
}
```
```js
  heading: {
    render: Heading,
    children: ['inline'],
    attributes: {
      id: { type: String },
      level: { type: Number, required: true, default: 1 },
      className: { type: String },
    },
    transform(node, config) {
      const attributes = node.transformAttributes(config)
      const children = node.transformChildren(config)
      const id = generateID(children, attributes)

      return new Tag(this.render, { ...attributes, id }, children)
    },
  },
  ```
  I tried just doing the following 
  
  
```js
  heading: {
    transform(node, config) {
      const attributes = node.transformAttributes(config)
      const children = node.transformChildren(config)
      const id = generateID(children, attributes)

      return new Tag(this.render, { ...attributes, id }, children)
    },
  },
``` 
But it still makes my right sidebar completely disappear. Is there any way to just change the id generation part and leave the rest intact/default as is?

**Reply by rpaul-stripe** [reply, marked as answer]
If you want a default heading that *just* adds the ID, your second example is close, but what you are missing is splatting in the default heading node configuration:

```
  heading: {
    ...Markdoc.nodes.heading,
    transform(node, config) {
      const attributes = node.transformAttributes(config)
      const children = node.transformChildren(config)
      const id = generateID(children, attributes)

      return new Tag(this.render, { ...attributes, id }, children)
    },
  },
```

And in order for that to work, you have to import markdoc:

```
import Markdoc from "@markdoc/markdoc";
```

But I'm not sure how that would be affecting your sidebar. And this is also going to result in headings rendering as normal HTML heading tags instead of using a React component like your first example. It's hard to troubleshoot what's going on exactly without more information about your setup.

**Comment by cryptobench** [comment]
@rpaul-stripe I found the issue. So when I define a custom Heading node and I try to console.log the node in my code, then it returns this. My sidebar expects the name h1, h2, h3, h4, h5 and h6 , but after defining this custom node, everything returns "Heading" instead of the previous h1, h2, h3, h4, h5 and h6 etc. 

Any idea what can be done to revert this? Once that is done, then im all good.

```json
{
    "$$mdtype": "Tag",
    "name": "Heading",
    "attributes": {
        "level": 2,
        "id": "functions"
    },
    "children": [
        "Functions"
    ]
}
```

**Reply by cryptobench** [reply]
Solved! Was missing the h attributes in the return Tag

```js
return new Tag(
        `h${node.attributes['level']}`,
        { ...attributes, id },
        ch
```

Thanks for the assistance! Much appreciated!
